src/snowflake/snowpark/async_job.py

```diff
@@ -0,0 +1,226 @@
+        - Temporary objects (e.g., tables) might be created when evaluating dataframes and they will be dropped automatically after all queries finish when calling a synchronous API. When you evaluating dataframes asynchronously, temporary objects will only be dropped after calling :meth:`result`.
+    """
+
+    def __init__(
+        self,
+        query_id: str,
+        query: str,
+        server_connection: "snowflake.snowpark._internal.server_connection.ServerConnection",
+        data_type: _AsyncDataType = _AsyncDataType.ROW,
+        **kwargs,
+    ) -> None:
+        #: The query ID of the executed query
+        self.query_id = query_id
+        #: The SQL text of of the executed query
+        self.query = query
+        self._conn = server_connection._conn
+        self._cursor = self._conn.cursor()
+        self._data_type = data_type
+        self._parameters = kwargs
+        self._result_meta = None
+        self._server_connection = server_connection
+        self._inserted = False
+        self._updated = False
+        self._deleted = False
+        self._plan = None
+
+    def is_done(self) -> bool:
+        """Check the status of the query associated with this instance and returns a bool value indicating whether the query has
+        finished.
+
+        """
+        status = self._conn.get_query_status(self.query_id)
+        is_running = self._conn.is_still_running(status)
+
+        return not is_running
+
+    def cancel(self) -> None:
+        """Cancels the query associated with this instance."""
+        # stop and cancel current query id
+        self._cursor.execute(f"select SYSTEM$CANCEL_QUERY('{self.query_id}')")
+
+    def _table_result(
+        self, result_data: Union[List[tuple], List[dict]]
+    ) -> Union[
+        "snowflake.snowpark.MergeResult",
+        "snowflake.snowpark.UpdateResult",
+        "snowflake.snowpark.DeleteResult",
+    ]:
+        if self._data_type == _AsyncDataType.UPDATE:
+            return snowflake.snowpark.UpdateResult(
+                int(result_data[0][0]), int(result_data[0][1])
+            )
+        elif self._data_type == _AsyncDataType.DELETE:
+            return snowflake.snowpark.DeleteResult(int(result_data[0][0]))
+        else:
+            idx = 0
+            rows_inserted, rows_updated, rows_deleted = 0, 0, 0
+            if self._inserted:
+                rows_inserted = int(result_data[0][idx])
+                idx += 1
+            if self._updated:
+                rows_updated = int(result_data[0][idx])
+                idx += 1
+            if self._deleted:
+                rows_deleted = int(result_data[0][idx])
+            return snowflake.snowpark.MergeResult(
+                rows_inserted, rows_updated, rows_deleted
+            )
+
+    def result(self) -> Union[List[Row], "pandas.DataFrame", Iterator[Row], int, None]:
```

Comment:
I'm wondering, can this method or part of the code be shared with some existing function? I imagine most of the process will be similar to how we process the result today, so we could avoid some code duplication.

Reply:
So actually this function is a bit different from what we have now. Most of other code is synchronously executed, so you dont have to use the query id to fetch results. Instead, you get a cursor instantly and processing the result is based on the cursor

Reply:
I see. I think my worry is mainly about if we need to change how we process the result, we might have to change two places. But it seems most of the actual processing is pushed down to connector, so we are probably okay.